src/textual_serve/download_manager.py

```diff
@@ -0,0 +1,175 @@
+from __future__ import annotations
+
+import asyncio
+from dataclasses import dataclass, field
+import logging
+from typing import AsyncGenerator, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from textual_serve.app_service import AppService
+
+log = logging.getLogger("textual-serve")
+
+DOWNLOAD_TIMEOUT = 4
+DOWNLOAD_CHUNK_SIZE = 1024 * 64  # 64 KB
+
+
+@dataclass
+class Download:
+    app_service: "AppService"
+    delivery_key: str
+    file_name: str
+    open_method: str
+    mime_type: str
+    encoding: str | None = None
+    """The encoding of the content.
+    Will be None if the content is binary.
+    """
+    incoming_chunks: asyncio.Queue[bytes | None] = field(default_factory=asyncio.Queue)
+
+
+class DownloadManager:
+    """Class which manages downloads for the server.
+
+    Serves as the link between the web server and app processes during downloads.
+
+    A single server has a single download manager, which manages all downloads for all
+    running app processes.
+    """
+
+    def __init__(self):
+        self._active_downloads: dict[str, Download] = {}
+        """A dictionary of active downloads.
+
+        When a delivery key is received in a meta packet, it is added to this set.
+        When the user hits the "/download/{key}" endpoint, we ensure the key is in
+        this set and start the download by requesting chunks from the app process.
+
+        When the download is complete, the app process sends a "deliver_file_end"
+        meta packet, and we remove the key from this set.
+        """
+
+    async def create_download(
+        self,
+        *,
+        app_service: "AppService",
+        delivery_key: str,
+        file_name: str,
+        open_method: str,
+        mime_type: str,
+        encoding: str | None = None,
+    ) -> None:
+        """Prepare for a new download.
+
+        Args:
+            app_service: The app service to start the download for.
+            delivery_key: The delivery key to start the download for.
+            file_name: The name of the file to download.
+            open_method: The method to open the file with.
+        """
+        self._active_downloads[delivery_key] = Download(
+            app_service,
+            delivery_key,
+            file_name,
+            open_method,
+            mime_type,
+            encoding,
+        )
+
+    async def download(self, delivery_key: str) -> AsyncGenerator[bytes, None]:
+        """Download a file from the given app service.
+
+        Args:
+            app_service: The app service to download from.
```

Comment:
This isn't in the signature. I'm assuming it has been removed since you wrote it.

Reply:
Removed